lib/utils/iterators.go

```diff
@@ -0,0 +1,121 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package utils
+
+import (
+	"context"
+	"errors"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/lib/utils/pagination"
+)
+
+// ErrStopIteration is value that signals to stop iteration from the caller injected function.
+var ErrStopIteration = errors.New("stop iteration")
+
+// ForEachOptions specifies options for ForEachResource.
+type ForEachOptions struct {
+	// PageSize is the number of items to fetch per page.
+	PageSize int
+}
+
+// ForEachOptionFunc is a function that sets an option on ForEachOptions.
+type ForEachOptionFunc func(*ForEachOptions)
+
+// WithPageSize sets the page size option.
+func WithPageSize(pageSize int) ForEachOptionFunc {
+	return func(opts *ForEachOptions) {
+		opts.PageSize = pageSize
+	}
+}
+
+// TokenLister is a function that lists resources with a page token.
+type TokenLister[T any] func(context.Context, int, string) ([]T, string, error)
+
+// ForEachResource iterates over resources.
+// Example:
+//
+//	count := 0
+//	err := ForEachResource(ctx, svc.ListAccessLists, func(acl accesslist.AccessList) error {
+//	   count++
+//	   return nil
+//	})
+//	if err != nil {
+//	   return trace.Wrap(err)
+//	}
+//	fmt.Printf("Total access lists: %v", count)
+func ForEachResource[T any](ctx context.Context, listFn TokenLister[T], fn func(T) error, opts ...ForEachOptionFunc) error {
+	var options ForEachOptions
+	for _, opt := range opts {
+		opt(&options)
+	}
```

Comment:
nit: set default pageSize if not provided with `WithPageSize` ? its usually `apidefaults.DefaultChunkSize` value used as default. 

should we also validate the pagesize provided with  `WithPageSize`?

Reply:
The 0 value is valid as page side. I it will mean that the server will use default page size and it is commonly used patter in our codebase for instance https://github.com/gravitational/teleport/blob/fa97b8f90df0a0437055fbc09fe1f5461a127d32/tool/tctl/common/acl_command.go#L121